packages/flutter/lib/src/services/platform_channel.dart

```diff
@@ -0,0 +1,198 @@
+  final String name;
+
+  /// The message codec used by this channel, not `null`.
+  final MethodCodec codec;
+
+  /// Invokes a [method] on this channel with the specified [arguments].
+  ///
+  /// Returns a [Future] which completes to one of the following:
+  ///
+  /// * a result (possibly `null`), on successful invocation;
+  /// * a [PlatformException], if the invocation failed in the platform plugin;
+  /// * a [FormatException], if encoding or decoding failed.
+  Future<dynamic> invokeMethod(String method, [dynamic arguments]) async {
+    assert(method != null);
+    return codec.decodeEnvelope(await PlatformMessages.sendBinary(
+        name,
+        codec.encodeMethodCall(method, arguments),
+    ));
+  }
+
+  /// Sets up a broadcast stream for receiving events on this channel.
+  ///
+  /// Returns a broadcast [Stream] which emits events to listeners as follows:
+  ///
+  /// * a decoded data event (possibly `null`) for each successful event
+  /// received from the platform plugin;
+  /// * an error event containing a [PlatformException] for each error event
+  /// received from the platform plugin;
+  /// * an error event containing a [FormatException] for each event received
+  /// where decoding fails;
+  /// * an error event containing a [PlatformException] or [FormatException]
+  /// whenever stream setup fails (stream setup is done only when listener
+  /// count changes from 0 to 1).
+  ///
+  /// Notes for platform plugin implementers:
+  ///
+  /// Plugins must expose methods named `listen` and `cancel` suitable for
+  /// invocations by [invokeMethod]. Both methods are invoked with the specified
+  /// [arguments].
+  ///
+  /// Following the semantics of broadcast streams, `listen` will be called as
+  /// the first listener registers with the returned stream, and `cancel` when
+  /// the last listener cancels its registration. This pattern may repeat
+  /// indefinitely. Platform plugins should consume no stream-related resources
+  /// while listener count is zero.
+  Stream<dynamic> receiveBroadcastStream([dynamic arguments]) {
+    StreamController<dynamic> controller;
+    controller = new StreamController<dynamic>.broadcast(
+        onListen: () async {
+          PlatformMessages.setBinaryMessageHandler(
+              name,
+                  (ByteData reply) async {
+                if (reply == null) {
+                  controller.close();
+                } else {
+                  try {
+                    controller.add(codec.decodeEnvelope(reply));
+                  } catch (e) {
+                    controller.addError(e);
+                  }
+                }
+              }
+          );
+          try {
+            await invokeMethod('listen', arguments);
+          } catch (e) {
+            PlatformMessages.setBinaryMessageHandler(name, null);
+            controller.addError(e);
+          }
+        }, onCancel: () async {
+      PlatformMessages.setBinaryMessageHandler(name, null);
```

Comment:
Argh - the indentation is completely garbled in this class! Appears my IntelliJ does that on copy/paste. I moved nicely formatted code from PlatformMessages...